### DISCUSSION TITLE
The Keycloak 24.0.4 crash was caused by a constant increase in memory usage

### DISCUSSION THREAD

**Discussion by uwevil** [keycloak/keycloak#35245] [Archive]
Context: 

We have noticed that our Keycloak instance in production (consisting of 6 pods in OpenShift) is constantly crashing after some time due to memory issues. After taking a heap dump, we discovered that there were 4000 instances of PoolingHttpClientConnectionManager objects and we suspect that HTTP connections are not being closed after creation.

![image (1)](https://github.com/user-attachments/assets/ce7c0330-8e23-42b3-9f4a-accd032ae58f)
<img width="1596" alt="Screenshot 2024-11-22 at 16 03 37" src="https://github.com/user-attachments/assets/95951669-2b65-4c56-bfa8-845e77a4f228">

Most of the connections are timed-waiting (around 50%) and are continuously increasing over a period of three hours.

After examining the Keycloak code, we found the DefaultHttpClientFactory where HTTP configurations are set. Do you have any suggestions for how we can resolve this problem?

in **src/main/java/org/keycloak/connections/httpclient/DefaultHttpClientFactory.java**

```
long socketTimeout = config.getLong("socket-timeout-millis", 5000L);
                    long establishConnectionTimeout = config.getLong("establish-connection-timeout-millis", -1L);
                    int maxPooledPerRoute = config.getInt("max-pooled-per-route", 64);
                    int connectionPoolSize = config.getInt("connection-pool-size", 128);
                    long connectionTTL = config.getLong("connection-ttl-millis", -1L);
                    boolean reuseConnections = config.getBoolean("reuse-connections", true);
                    long maxConnectionIdleTime = config.getLong("max-connection-idle-time-millis", 900000L);
                    boolean disableCookies = config.getBoolean("disable-cookies", true);
                    String clientKeystore = config.get("client-keystore");
                    String clientKeystorePassword = config.get("client-keystore-password");
                    String clientPrivateKeyPassword = config.get("client-key-password");
                    boolean disableTrustManager = config.getBoolean("disable-trust-manager", false);
```

**Comment by jonkoops** [comment]
There have been a lot of improvements in this regard in more recent versions of Keycloak. Upgrading your Keycloak instance to the latest version might help resolve these problems. Either way, 26.x is currently the only version that receives active support (including vulnerabilities), so I would recommend you upgrade ASAP.
